Спринт 22.

Kris-found opened the pull request:
- Создан экран Новый плейлист
- Реализован экран Плейлисты
- Реализована логика добавления трека в плейлисты, создания плейлистов, отображение созданных плейлистов
- Изменения на экране Аудиоплейера: вместо названия альбома в заголовке теперь название трека
- На экране поиска курсор сниний (наверно)

Kris-found committed:
Спринт 22.
- Создан экран Новый плейлист
- Реализован экран Плейлисты
- Реализована логика добавления трека в плейлисты, создания плейлистов, отображение созданных плейлистов
- Изменения на экране Аудиоплейера: вместо названия альбома в заголовке теперь название трека
- На экране поиска курсор сниний (наверно)

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/Utils/Converters.kt:
```diff
@@ -12,6 +15,16 @@ fun dpToPx(dp: Float, context: Context): Int {
     ).toInt()
 }
 
+fun formattedTracksCount(count: Int): String {
```
Воспользуйся лучше plurals, в андроиде уже реализована логика, с помощью которой для каждой локали будет подбираться соответствующее окончание. Почитать о том, как это использовать, можно в официальной документации - https://developer.android.com/guide/topics/resources/string-resource#Plurals

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/library/data/PlaylistRepositoryImpl.kt:
```diff
@@ -0,0 +1,86 @@
+import java.io.FileOutputStream
+
+class PlaylistRepositoryImpl(
+    private val appDatabase: AppDatabase,
```
Не стоит сюда инжектить всю БД, лучше всегда инжектить только те ДАО интерфейсы, которые нужны в классе - это повысит тестируемость класса

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/library/data/PlaylistRepositoryImpl.kt:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    companion object {
+        const val CHILD_FILE_PATH = "Album"
```
Константу можно сделать приватной

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/library/data/PlaylistRepositoryImpl.kt:
```diff
@@ -0,0 +1,86 @@
+            val fileName = "cover_${System.currentTimeMillis()}.jpg"
+            val file = File(filePath, fileName)
+
+            val bitmap = Glide.with(app)
+                .asBitmap()
+                .load(uri)
+                .submit()
+                .get()
```
Использование глайда здесь излишне, так как битмап можно раздекодить из Uri с помощью `ContentResolver` и `BitmapFactory` - в таком случае будет выполнено гораздо меньше операций

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/library/data/db/entity/PlaylistEntity.kt:
```diff
@@ -0,0 +1,18 @@
+    val name: String,
+    val description: String,
+    val imageUri: String,
+    val tracksId: String,
```
Хранить лучше не в формате строки, а сделать связь между таблицами one-to-many, в документации есть пример - https://developer.android.com/training/data-storage/room/relationships/one-to-many

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/library/ui/favorite/FavoriteTracksFragment.kt:
```diff
@@ -72,9 +74,12 @@ class FavoriteTracksFragment : Fragment() {
-        val audioPlayerIntent = Intent(requireContext(), AudioPlayerActivity::class.java).apply {
-            putExtra(AudioPlayerActivity.KEY_TRACK_TAP, track)
+        val bundle = Bundle().apply {
+            putSerializable(AudioPlayerFragment.KEY_TRACK_TAP, track)
```
Для передачи данных между компонентами Android (активити, фрагменты, сервисы) лучше использовать интерфейс `Parcelable`. Для генерации реализации этого интерфейса есть плагин `kotlin-parcelize`. Чтобы реализовать нужно сделать следующее - добавить плагин в проект, доменную модель трека наследовать от интерфейса `Parcelable` и воспользоваться аннотацией `@Parcelize`. Ну и передавать сами данные не в формате String, а сразу в формате Parcelable. Тут есть пример работы с плагином - https://developer.android.com/kotlin/parcelize

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/library/ui/playlist/CreatePlaylistFragment.kt:
```diff
@@ -0,0 +1,156 @@
+
+    private val viewModel by viewModel<PlaylistsViewModel>()
+
+    private var savedImagePath: String? = null
```
Данные не рекомендуется хранить во фрагменте - для этого есть вьюмодель и её стейт

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/player/model/AddingTrackState.kt:
```diff
@@ -0,0 +1,8 @@
+package com.practicum.playlistmaker.player.model
+
+sealed class AddingTrackState {
```
Здесь больше подойдёт `sealed interface`, т.к. преимущества класса здесь не используются

Kris-found committed:
Спринт 22.
- устранен баг с вылетом при создании плейлиста
- улучшен код в соотвествии с некоторыми рекомендациями
